### PR TITLE
Refactor and extract common fields from payment app schemas

### PR DESCRIPTION
**Pull request by wikiti**
### WHY are these changes introduced?

To keep code maintanable, after creating the schemas for all payment app types, we decided to do a minor refactor to extract common fields into a base schema.

### WHAT is this pull request doing?

A new base schema, `BasePaymentsAppExtensionSchema`, is introduced with all the common fields. All existing schemas are going to extend this base schema, and add new fields when required.

Instead of extracting **strict** common fields (same types, same validations), I decided to leverage on the `omit` and `required` zod methods. This way, we can define *almost-common* fields, and make minor tweaks based on the payment app. For example, if a field is common and optional to all payment app types, but it is required within a specific type, then the `required` method would change the optional field from the base schema to be required in the sub schema. Something similar applies to `omit`; if a field is present in *almost* all payment app types, but not in all of them, the `omit` method can be used to skip that field for the specific type.

I tried to find a balance between extracting common fields and avoiding repetition. I think the current implementation doesn't create too much indirection between the base schema and the subschemas, but please let me know in comments if you find it hard to follow 🙏🏿 

I also made some minor changes to some schemas that were not properly formed, such as the custom credit card encryption certificate field

### How to test your changes?

- Create app with `npm init @shopify/app@latest`
- Pull this branch into your local `cli` copy
- Run `SHOPIFY_SERVICE_ENV=spin SPIN_INSTANCE={YOUR-SPIN-INSTANCE} NODE_TLS_REJECT_UNAUTHORIZED=0 SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 pnpm shopify app config link --path {YOUR-APP-PATH}`
- Run `SHOPIFY_SERVICE_ENV=spin SPIN_INSTANCE={YOUR-SPIN-INSTANCE} NODE_TLS_REJECT_UNAUTHORIZED=0 SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 pnpm shopify app deploy --path {YOUR-APP-PATH}`
- View `Payments Extensions`

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/credit_card_payments_app_extension_schema.ts [hunk 1]
```diff
@@ -1,52 +1,38 @@
...
+
+export const CreditCardPaymentsAppExtensionSchema = BasePaymentsAppExtensionSchema.required({
+  refund_session_url: true,
+  capture_session_url: true,
+  void_session_url: true,
 })
+  .omit({
+    buyer_label: true,
+    buyer_label_translations: true,
+  })
```
Comment (wikiti) [inline packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/credit_card_payments_app_extension_schema.ts:18]: After a brief discussion with @isaacroldan, I'll change the usage of `omit` to instead use composed schemas (e.g. `BasePaymentsAppExtensionSchema.extend(BuyerLabelSchema).extend({....})`)

File: packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/base_payments_app_extension_schema.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import {BaseSchema} from '../../schemas.js'
+import {zod} from '@shopify/cli-kit/node/schema'
+
+const MAX_LABEL_SIZE = 50
+
+export const BasePaymentsAppExtensionSchema = BaseSchema.extend({
```
Comment (marklevi) [inline packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/base_payments_app_extension_schema.ts:6]: I like the approach of using composition to build each schema using logical groupings like `DeferredPaymentsSchema` and `ConfirmationSchema`. However, I am not so sure if lumping a group of fields under `BasePaymentsAppExtensionSchema` in aligned with that idea. I think it's fine to have some duplicate fields in each extension schema if there isnt a logical grouping associated with those. What do you think?

Comment (wikiti) [reply packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/base_payments_app_extension_schema.ts:6]: > I think it's fine to have some duplicate fields in each extension schema if there isnt a logical grouping associated with those. Nonetheless, it isnt a sticking point. What do you think?

That breaks the idea of lifting common fields into a single schema. Having schemas being defined as `BaseSchema.merge(PaymentsAppUrlExtensionsSchema).merge(MerchantLabelSchema).merge(InputSchema).merge(TestModeSchema)` seems insane to me. That's a lot of boilerplate code for just strictly sticking to composition.

I think we get the best of both worlds by using a common schema to all payment apps (`BasePaymentsAppExtensionSchema`), and then, to prevent code duplication, using schema "plugins" to extend the available fields. 

It's what we already do for the common `BaseSchema` schema, common to all extensions

**Review by marklevi — APPROVED** [review]

File: packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/base_payments_app_extension_schema.ts (hunk 2 above)
Comment (marklevi) [reply packages/app/src/cli/models/extensions/specifications/payments_app_extension_schemas/base_payments_app_extension_schema.ts:6]: I think the extracting out the common properties has come at a slight cost in terms of code readability but I understand the approach. approving the PR !
